<h1>INTRODUCTIONS</h1>
<h2>Live Code 2 </h2>
<h3><b>Muhamad Natual Hisak</b></h3>

<h2>PROBLEM 1</h2>
Kamu sedang mencari rumah dan akan memilih satu dari dua pilihan rumah dengan harga yang cukup berbeda. Sekilas, kedua rumah sama karakteristiknya, namun kamu ingin memastikan kembali apakah rumah yang lebih mahal patut dibeli atau tidak.<br>


Keterangan:

-Pada kolom Jumlah Kamar Tidur, Jumlah Kamar Mandi, Ada Halaman Belakang?, Ada Parkir Mobil dan Motor?, serta Dekat Dengan Pusat Kota? bernilai biner yaitu 1: Ya dan 0: Tidak.<br>

-Kolom harga dalam satuan ratus juta.<br>

-Buatlah vektor yang merupakan representasi masing-masing rumah berdasarkan tabel di atas dengan mengecualikan kolom harga dan hitung cosine similarity antar kedua vektor. Kemudian jawab pertanyaan berikut di markdown:<br>

a. Apakah kedua rumah mirip? jika iya, mengapa dan jika tidak, mengapa (jawab berdasarkan hasil perhitungan cosine similaritynya)? Dan keputusan yang akhirnya kamu pilih, rumah 1 atau rumah 2? apa pertimbangannya?

b. Jika meninjau dua buah vektor dan dihitung cosine similaritynya (cos theta), jelaskan secara singkat, jelas, padat apa makna cosine similarity yang bernilai 0 dan 1 (tinjau dari posisi dua vektor di koordinat kartesian)?

c. Mengapa cosine similarity harus melibatkan vektor bukan matriks?


<h2>JAWABAN PROBLEM 1</h2>

import library yang akan digunakan yaitu numpy, matplotlib, dan pandas

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

Jadikan data kedua rumah kedalam vektor

In [2]:
Rumah1 = np.array([100,240,3,3,2,0,1,1])
Rumah2 = np.array([120,200,2,2,2,1,1,1])
Rumah1, Rumah2

(array([100, 240,   3,   3,   2,   0,   1,   1]),
 array([120, 200,   2,   2,   2,   1,   1,   1]))

Hitung nilai dot product dari kedua vektor

In [3]:
dot = np.dot(Rumah1,Rumah2, out = None)
print("Nilai dot product antara kedua vektor : ", dot)

Nilai dot product antara kedua vektor :  60018


Hitung nilai Cosine Similarity

In [4]:
Cosim = dot/(np.linalg.norm(Rumah1)*(np.linalg.norm(Rumah2)))
print("Cosine similarity antara kedua vektor : ",Cosim)

Cosine similarity antara kedua vektor :  0.9893997002438046


Jawaban dari pertanyaan :

a. Kedua rumah mirip karena, berdasarkan nilai cosine similarity antara kedua rumah yang bernilai 98.9%, itu artinya kedua rumah sangat mirip karena nilai cosine similarity menunjukkan kemiripan antar 2 buah vektor.<br>
Rumah yang saya pilih adalah rumah 2 karena dengan harga yang lebih murah, saya bisa mendapatkan rumah dengan luas tanah yang lebih besar, dan rumah yang memiliki halaman belakang walaupun luas bangunan, jumlah lantai dan kamar tidur yang tidak sebanyakan rumah 1.

b. Jika meninjau dua buah vektor dan dihitung cosine similaritynya, nilai cosine similarity 0 mengartikan bahwa sudut yang terbentuk dari kedua vektor hampir mencapai 90 derajat, yang berarti tidak terjadi kemiripan antar 2 buah vektor tersebut. Sedangkan nilai cosine similarity 1 mengartikan bahwa sudut yang terbentuk dari kedua vektor hampir mencapai 0 derajat, yang berarti kedua vektor tersebut sangat mirip.

c. Cosine similarity harus melibatkan vektor bukan matrix karena inti dari perhitungan cosine similarity ini adalah menghitung sudut yang terbentuk dari dua buah vektor. Jika digambar pada diagram kartesian, sudut tersebut akan terbentuk dari dua buah garis yang ditarik berdasarkan nilai vektor tersebut.

<h2>PROBLEM 2</h2>
Kamu diberikan tugas untuk menganalisis harga sewa apartemen di Argentina per Januari 2015. Gunakan kemampuan statistikamu untuk menganalisis data yang tersedia!<br>

Pada graded challenge ini, data diakses menggunakan bigquery-public-data pada Google Cloud Big Query.
Buka Google Cloud Platform, masuk ke BigQuery, lalu buka tab bigquery-public-data atau klik link berikut atau link berikut untuk langsung menuju ke tabel.<br>

1. Gunakan tabel properties_rent_201501 pada dataset properati_properties_ar.
2. Buatlah query dengan kriteria sebagai berikut:
3. Pilih HANYA kolom price
4. Pilih data dengan value operation 'rent' dan property_type 'Apartement`
5. Simpan dataset dalam bentuk csv, dengan nama h8dsft_P0LC3_<nama-students>.csv.
6. Salin query yang telah dibuat di Google Cloud Platform, tulislah pada bagian atas notebook!
7. Tampilkan head dan tail dari dataset pada notebook!

<b>Clue</b>: Untuk mengetahui adanya anomali, kamu bisa menggunakan metode extreme value analysis. Untuk melakukan pengecekan anomali/outlier, lakukan langkah-langkah di bawah ini:<br>

1. Lakukan perhitungan central tendency (mean, median, modus) terhadap data sebelum dideteksi adanya anomali.
2. Cek skewness data untuk mengetahui apakah data terdistribusi normal atau tidak.
3. Lakukan pengolahan data dengan menggunakan extreme value analysis.
4. Buat variabel baru yang menyimpan data yang sudah dibuang data anomalinya.


<h3>PERTANYAAN</h3>

Silahkan jawab pertanyaan-pertanyaan di bawah ini berdasarkan hasil yang kamu peroleh dan tulis pada bagian hasil:

1. Berapa rata-rata, median, dan modus dari data tersebut sebelum dihilangkan outliernya? Bagaimana kecerendungan pemusatan datanya? jelaskan jawabanmu!
2. Sebelum melakukan extreme value analysis, kamu harus melakukan pengecekan skewness dari distribusi datanya. Apakah datanya skew atau normal? jelaskan jawabanmu!
3. Ada dua teknik untuk melakukan extreme value analysis, teknik yang mana yang kamu pakai? berikan alasanmu berdasarkan data!

<h2>JAWABAN PROBLEM 2</h2>

Lakukan query pada BigQuery dengan kode dibawah : <br>

<b>SELECT price FROM bigquery-public-data.properati_properties_ar.properties_rent_201501 where property_type = "apartment" and operation = "rent"</b><br>

Setelah itu save dataset agar bisa di load ke program yang akan dibuat

Load dataset yang sudah di download dari bigquery, lalu tampilkan

In [5]:
df = pd.read_csv(r"C:\Users\Omen\Desktop\Hacktiv8\Live Code 2\h8dsft_P0LC3_Muhamad_Natual_Hisak.csv.csv")
df

price
0      2500
1      3500
2      2850
3      1850
4      1000
...     ...
1653    500
1654    500
1655    500
1656    500
1657    500

[1658 rows x 1 columns]

In [6]:
df.head() #menampilkan 5 data teratas

price
0   2500
1   3500
2   2850
3   1850
4   1000

In [7]:
df.tail()#menampilkan 5 data terbawah

price
1653    500
1654    500
1655    500
1656    500
1657    500

Copy data ke variabel baru agar data asli tidak ter-overwrite saat dilakukan proses cleaning

In [8]:
price = df.copy()
price.head()

price
0   2500
1   3500
2   2850
3   1850
4   1000

Cek missing value

In [9]:
price.info(null_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1658 entries, 0 to 1657
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   price   1658 non-null   int64
dtypes: int64(1)
memory usage: 13.1 KB


C:\Users\Omen\AppData\Roaming\Python\Python37\site-packages\ipykernel_launcher.py:1: FutureWarning: null_counts is deprecated. Use show_counts instead
  """Entry point for launching an IPython kernel.


Dari data diatas bisa dilihat bahwa tidak ada missing value dari tabel

Menghitung Central Tendency (Mean, Median, Modus) pada data sebelum dilakukan cleaning

In [10]:
Mean = price['price'].mean()
Median = price['price'].median()
Mode = price['price'].mode()

print("mean : ", Mean)
print("median : ", Median)
print("modus : ", Mode)

mean :  5166.355247285886
median :  2700.0
modus :  0    0
dtype: int64


Mencari jumlah kemunculan data 0 sebagai modus pada kolom price 

In [11]:
price.loc[price.price == 0]

price
1193      0
1194      0
1195      0
1196      0
1197      0
...     ...
1641      0
1642      0
1643      0
1644      0
1645      0

[453 rows x 1 columns]

In [12]:
(453/1658)*100

27.322074788902288

Bisa dilihat bahwa nilai 0 muncul sebanyak 453 kali dalam data ini atau sebanyak 27.3% input pada data ini adalah bernilai 0

Cek distribusi dengan mencari nilai skewness data

In [13]:
skew = price["price"].skew()
print("skewness : ",skew)

skewness :  4.2401362864090695


Dari data skewness diatas, bisa disimpulkan bahwa data tersebut memiliki distribusi sangat tidak normal atau bisa disebut highly skewed. Hal ini bisa dilihat dari nilai skewness yang berada diatas 1.

Dikarenakan data tersebut termasuk highly skewed (skew > 1 OR skew < -1), maka untuk extreme value analysis akan digunakan metode IQR dengan upper limit bernilai Q3+3 * IQR dan lower limit Q1-3 * IQR<br>

Untuk melakukan hal tersebut, kita butuh nilai Q1 Q3 dan IQR. Selanjutnya mencari ketiga nilai tersebut.

In [14]:
Q1 = price["price"].quantile(.25)
Q3 = price["price"].quantile(.75)
IQR = Q3-Q1
print("Q1 :",Q1)
print("Q3 :",Q3)
print("IQR :",IQR)

Q1 : 0.0
Q3 : 5500.0
IQR : 5500.0


Selanjutnya menentukan limit bawah dan atas

In [15]:
low = Q1-3*IQR
high = Q3+3*IQR
print("lower limit",low)
print("upper limit",high)

lower limit -16500.0
upper limit 22000.0


Selanjutnya mencari data data yang termasuk dalam outlier (price > upper limit OR price < lower limit)

In [16]:
out = price.loc[(price.price >= high) | (price.price <= low)]
out

price
27   30000
33   45000
130  23000
135  80000
137  46000
..     ...
806  28000
847  24000
898  76000
928  25000
952  24000

[82 rows x 1 columns]

In [17]:
(82/1658)*100

4.945717732207479

Ada 82 buah data yang termasuk outlier dari total 1658 data, yang berarti persentase outlier pada data ini ada di angka 4.946%<br>

Selanjutnya hapus data yang termasuk dalam outlier, lalu tabel baru yang sudah bersih dari outlier dimasukkan ke variabel baru

In [18]:
price_clean = price.loc[(price.price < high) & (price.price > low)]
price_clean

price
0      2500
1      3500
2      2850
3      1850
4      1000
...     ...
1653    500
1654    500
1655    500
1656    500
1657    500

[1576 rows x 1 columns]

Tabel diatas merupakan tabel price yang sudah dibersihkan anomali data nya.

Jawaban dari pertanyaan : 
1. Sebelum dihilangkan outlier nya, data tersebut memiliki: <br>
    a. nilai rata-rata : 5166.355<br>
    b. median :  2700.0<br>
    c. modus :    0<br>
    dengan nilai mean(rata-rata) yang lebih besar dari median, menandakan bahwa banyaknya data yang lebih besar dari median yang kemungkinan bisa dianggap sebagai outlier, atau bisa juga data yang nilainya lebih besar dari median berjumlah banyak. Modus yang bernilai 0 dan muncul sebanyak 453 input juga menandakan bahwa ada banyak data yang bernilai jauh lebih besar dari median karena rata-rata yang didapat juga bernilai sangat besar yaitu 5166.355

    
2.  Persebaran datanya tidak normal atau highly skewed, hal ini bisa dilihat dari nilai skew yang didapatkan yaitu : 4.240.<br>
    Data lebih tersebar padat di sebelah kiri sehingga data bisa disebut right-skewed

3. Teknik yang dipakai untuk melakukan extreme value analysis untuk data ini adalah dengan memanfaatkan Tukey's rule atau menggunakan IQR. Hal ini saya lakukan karena bentuk persebaran data yang tidak normal atau highly skewed sehingga metode tersebut cocok untuk digunakan.
